src/tape.jl

```diff
@@ -0,0 +1,171 @@
+using MacroTools
+import MacroTools: splitdef, combinedef
+
+mutable struct TapeNode
+    op::Function
+    args::Vector
+    results::Vector
+    kwargs::Dict
+end
+
+TapeNode(op, args, results; kwargs...) =  TapeNode(op, args, results, kwargs)
+
+mutable struct Tape
+    nodes::Dict{TensorHandle, TapeNode}
+    attrs::Dict
+end
+
+Tape(;kwargs...) = Tape(Dict{TensorHandle, TapeNode}(), Dict(kwargs...))
+
+function set_tape(new_tape=nothing)
+    if new_tape === nothing
+        new_tape = Tape()
+    end
+    context = Context()
+    context.attrs["tape"] = new_tape
+    push!(global_context, context)
+    return new_tape
+end
+
+function get_tape()
+    return global_context["tape"]
+end
+
+function add_node(t, node)
+    tape = get_tape()
+    tape === nothing && return
+    tape.nodes[t] = node
+end
+
+grad_fns = Dict()
+
+macro back_for(target, fn)
```

Comment:
This should be be changed to instread make function methods
```
backwards(::typeof(Ops.add))= function(grad, x, y; kwargs...)
    println("Add got $grad, $x, $y")
    return [constant(1.0).*grad, constant(1.0).*grad]
end)
```
etc.

Reply:
done